arch/arm/crc32_armv8.c

```diff
@@ -9,78 +9,106 @@
         }
 
-        if ((len >= sizeof(uint16_t)) && ((ptrdiff_t)buf & (sizeof(uint32_t) - 1))) {
-            buf2 = *((uint16_t*)buf);
-            c = __crc32h(c, buf2);
-            buf += sizeof(uint16_t);
+        if ((len >= sizeof(uint16_t)) && ((ptrdiff_t)src & (sizeof(uint32_t) - 1))) {
+            val2 = *((uint16_t*)src);
+            if (COPY) {
+                *((uint16_t*)dst) = val2;
+                dst += sizeof(uint16_t);
+            }
+            c = __crc32h(c, val2);
+            src += sizeof(uint16_t);
             len -= sizeof(uint16_t);
         }
 
-        if ((len >= sizeof(uint32_t)) && ((ptrdiff_t)buf & (sizeof(uint64_t) - 1))) {
-            buf4 = *((uint32_t*)buf);
-            c = __crc32w(c, buf4);
+        if ((len >= sizeof(uint32_t)) && ((ptrdiff_t)src & (sizeof(uint64_t) - 1))) {
+            val4 = *((uint32_t*)src);
+            if (COPY) {
+                *((uint32_t*)dst) = val4;
+                dst += sizeof(uint32_t);
+            }
+            c = __crc32w(c, val4);
+            src += sizeof(uint32_t);
             len -= sizeof(uint32_t);
-            buf += sizeof(uint32_t);
         }
-
     }
 
     while (len >= sizeof(uint64_t)) {
-        buf8 = *((uint64_t*)buf);
-        c = __crc32d(c, buf8);
+        val8 = *((uint64_t*)src);
+        if (COPY) {
+            *((uint64_t*)dst) = val8;
+            dst += sizeof(uint64_t);
+        }
+        c = __crc32d(c, val8);
+        src += sizeof(uint64_t);
         len -= sizeof(uint64_t);
-        buf += sizeof(uint64_t);
     }
 
     if (len & sizeof(uint32_t)) {
-        buf4 = *((uint32_t*)buf);
-        c = __crc32w(c, buf4);
-        buf += sizeof(uint32_t);
+        val4 = *((uint32_t*)src);
+        if (COPY) {
+            *((uint32_t*)dst) = val4;
+            dst += sizeof(uint32_t);
+        }
+        c = __crc32w(c, val4);
+        src += sizeof(uint32_t);
     }
 
     if (len & sizeof(uint16_t)) {
-        buf2 = *((uint16_t*)buf);
-        c = __crc32h(c, buf2);
-        buf += sizeof(uint16_t);
+        val2 = *((uint16_t*)src);
+        if (COPY) {
+            *((uint16_t*)dst) = val2;
+            dst += sizeof(uint16_t);
+        }
+        c = __crc32h(c, val2);
+        src += sizeof(uint16_t);
     }
 
     if (len & sizeof(uint8_t)) {
-        c = __crc32b(c, *buf);
+        if (COPY)
+            *dst = *src;
+        c = __crc32b(c, *src);
     }
 
     c = ~c;
     return c;
 }
 
+Z_INTERNAL Z_TARGET_CRC uint32_t crc32_armv8(uint32_t crc, const uint8_t *src, size_t len) {
+    return crc32_armv8_copy_impl(crc, NULL, src, len, 0);
+}
+
 /* Note: Based on generic crc32_fold_* implementation with functable call replaced by direct call. */
 Z_INTERNAL Z_TARGET_CRC void crc32_fold_copy_armv8(crc32_fold *crc, uint8_t *dst, const uint8_t *src, size_t len) {
-    crc->value = crc32_armv8(crc->value, src, len);
-    memcpy(dst, src, len);
+    crc->value = crc32_armv8_copy_impl(crc->value, dst, src, len, 1);
```

Comment:
Depending on benchmark results when building with `-mno-unaligned-access`, it may still be faster to use `memcpy` on platforms that require it rather than writing to the source buffer one byte at a time:

```suggestion
#if !(defined(__ARM_FEATURE_UNALIGNED) || defined(_WIN32))
    if (((uintptr_t)src & (sizeof(uint64_t) - 1)) != ((uintptr_t)dst & (sizeof(uint64_t) - 1)) {
        crc->value = crc32_armv8_copy_impl(crc->value, NULL, src, len, 0);
        memcpy(dst, src, len);
    } else
#else
    {
        crc->value = crc32_armv8_copy_impl(crc->value, dst, src, len, 1);
    }
```

Reply:
That's interesting.